druid-shell/src/backend/x11/window.rs

```diff
@@ -909,8 +909,8 @@ impl Window {
     }
 
     fn add_invalid_rect(&self, rect: Rect) -> Result<(), Error> {
-        // expanding not needed here, because we are expanding at every use of invalid
-        borrow_mut!(self.invalid)?.add_rect(rect);
+        let scale = self.scale.get();
+        borrow_mut!(self.invalid)?.add_rect(rect.to_px(scale).expand().to_dp(scale));
```

Comment:
So clearly the comment was not correct (maybe the missing expand was the clipping stuff at the beginning of `render`?), but I'm not sure if this fix is correct either, because it might end up double-expanding in some situations.

Reply:
yeah the comment wasn't correct. there should be expand at https://github.com/linebender/druid/blob/fdc7562483188210971e13d71b1d400885ad9af6/druid-shell/src/backend/x11/window.rs#L770-L773 but still we will pass unexpanded invalid to druid https://github.com/linebender/druid/blob/fdc7562483188210971e13d71b1d400885ad9af6/druid-shell/src/backend/x11/window.rs#L786

now add_invalid_rect and render is same as gtk backend. 

> double-expanding in some situations

example?

Reply:
There's an `expand` in `PresentData::present`, so with this change, the rectangle that gets presented gets converted to px, expanded, converted back to dp, converted back to px, and expanded again. Since there might be numerical error in the conversion-to-dp-and-back-to-px, the final presented rectangle could be larger than the one used while clipping the cairo context.

Reply:
because before conversion-to-dp-and-back-to-px, it was aligned to px, `round` should fix the numerical errors. This is done by windows backend

Reply:
Yup, as long as there's only one `expand` and the rest are `round`s, it should be fine.